npeps/npep-173-service-account-matches.md

```diff
@@ -0,0 +1,51 @@
+# NPEP-173: NPEP template
+
+* Issue: [#173](https://github.com/kubernetes-sigs/network-policy-api/issues/173)
+* Status: Provisional
+
+## TLDR
+
+Ability to express subjects and peers in (Baseline)AdminNetworkPolicy API
+using [Kubernetes service accounts](https://kubernetes.io/docs/concepts/security/service-accounts/).
+
+## Goals
```

Comment:
I think that the main goal is to provide network policies based on identities, and service accounts are a good primitive in Kubernetes to do that.
I also think that we want to look for this new model to be more scalable to avoid implementation having to track all the existing Pod IPs

@thockin your thoughts?

Reply:
I don't understand the interest in describing this as being "based on identities", particularly if you say that arbitrary CIDRs can be "identities". How is calling this "based on identities" better than saying "you can select CIDRs or service accounts or pod labels"?

Reply:
yea I have added the scale and consistency user story to the next section under user stories, maybe I can change the goals to reflect "identity" or refactor the way I am writing "service account as a way to select pods"

Depending on whether we feel this is a new API CRD or part of ANP -> in which case we need to decide also the points Dan is raising:  https://github.com/kubernetes-sigs/network-policy-api/pull/274/files#r1850516386 we might need to tweak the TLDR/Goals sections here

Reply:
I second mike here. The main benefits I can see are;

1. Users dont interact with CIDRs/IPs.
2. _Some_ implementations of the API dont need to track IPs to provide this functionality (when using identities). 
3. Serviceaccounts are more secure than labels. e.g they dont span namespaces, can not be changed at runtime. 

We dont have to be opinionated on how the API would be implemented, and `service account as a way to select pods` can just be an implementation detail.

Reply:
My comments were based on Cilium's definitions, where (AIUI) everything you can apply a network policy toward is considered an "identity". (e.g., [this blog post](https://isovalent.com/blog/post/intro-to-cilium-network-policies/#decoupling-identity-from-ip-addresses-cilium-endpoint-and-identity) talks about how it turns podSelectors into "identities").

I had initially assumed people were talking about the same thing here. In particular, I had assumed that people were still talking about having a mix of "traditional" NetworkPolicy rules and new service-account-based NetworkPolicy rules, but calling the whole thing "identity-based", which seemed pointless.

But it sounds like what most of the people here actually want is a separate parallel network policy stack, which _only_ uses "identity-based" rules based on service-account/mTLS/SPIFFE/whatever.

Reply:
> My comments were based on Cilium's definitions, where (AIUI) everything you can apply a network policy toward is considered an "identity". (e.g., [this blog post](https://isovalent.com/blog/post/intro-to-cilium-network-policies/#decoupling-identity-from-ip-addresses-cilium-endpoint-and-identity) talks about how it turns podSelectors into "identities").
> 
> I

cilium identities are clear a no goal, and now I realize that makes the terminology confusing.

If we consider identity as "the set of characteristics that uniquely identify an application (running in a Pod)".
IPs can uniquely identify a Pod, but if a deployment is rolled out the IPs will change (pods don't restart are recreated), so the rules applied directly to IPs will not be identity based.

We can use the labels to identify the pods of the deployments/daemonsets but this identity system seems mostly based on the implementation details of kubernetes, and labels can be added arbitrarily. 

[Service Accounts](https://kubernetes.io/docs/concepts/security/service-accounts/) by the kubernetes docs definitions are "a type of non-human account that, in Kubernetes, provides a distinct identity in a Kubernetes cluster. Application Pods, system components, and entities inside and outside the cluster can use a specific ServiceAccount's credentials to identify as that ServiceAccount. "

I think that ServiceAccounts can work independently on how we translate the API to the implementation, map ServiceAccount to IPs, use ServiceAccount Tokens or Certificates , ....  and also allow us to define cross-cluster or cross-services policies

Reply:
hmm my understanding was the same as Dan's which is why I called out the non-goal here but seems like that _is_ one of the main goals looks like :)

Reply:
we discussed this during yesterday's netpol api meeting and agreed that SPIFFEEID can be a non-goal for V0 maybe but maybe good to track the user stories for all of this first